### DISCUSSION TITLE
Search for matching brackets before adding one

### DISCUSSION THREAD

**Discussion by DoctorRainer** [helix-editor/helix#10025] [General]
For example, if we have:

```
type Person struct --> here <--
}
```

... typing { automatically makes it to be {} which doesn't make sense

**Comment by Abdillah** [comment]
Even on the same line, typing `{` also gives duplicate closing bracket currently.

```
<?php

class SomeClassName --> type here <--}
```

**Edit for clarification:**
I did mean, if the missing bracket and closing bracket has something in-between, it might fails short.
But, this is actually intentionally not yet handled.

**Comment by DoctorRainer** [comment]
I don't understand why it is intentional

**Comment by pascalkuthe** [comment]
its intentional because:
* autopairs should work in nested brackets. For example inside a function call you still want autopairs to work:`foo(bar<cursor>)` 
* figuring out which brackets are unclosed is highly ambiguous and language dependent and TS can't handle invalid syntax trees like that well. It's probably not possible to figure this out reliably for most languages
* predictability > minor convenience: its better autopairs behave predictably rather than making its behaviour unreliable and eratic

**Comment by Abdillah** [comment]
> It's probably not possible to figure this out reliably for most languages

I do still figuring out if the autopairs can be powered by tree-sitter. Other more mature editor has been able to handle better, so Helix should be.

Probably, we just don't insert the auto pair if on the next tree-sitter's ERROR within a smallest scope that contains current insertion and the closing bracket actually contains an orphaned closing pair.

**Reply by Abdillah** [reply]
Typescript is a bit problematic but actually possible:

![Screenshot from 2024-03-28 12-05-26](https://github.com/helix-editor/helix/assets/3030950/c3ba5bfc-df60-473b-bcc4-6e284a51dde4)
